src/ruby/ext/grpc/rb_compression_options.c

```diff
@@ -0,0 +1,351 @@
+  long name_len = 0;
+  VALUE ruby_str = Qnil;
+
+  TypedData_Get_Struct(self, grpc_rb_compression_options,
+                       &grpc_rb_compression_options_data_type, wrapper);
+
+  /* Take both string and symbol parameters */
+  ruby_str = rb_funcall(new_level, rb_intern("to_s"), 0);
+
+  level_name = RSTRING_PTR(ruby_str);
+  name_len = RSTRING_LEN(ruby_str);
+
+  /* Check the compression level of the name passed in, and see which macro
+   * from the GRPC core header files match. */
+  if (strncmp(level_name, "none", name_len) == 0) {
+    grpc_rb_compression_options_set_default_level_helper(
+        wrapper->wrapped, GRPC_COMPRESS_LEVEL_NONE);
+  } else if (strncmp(level_name, "low", name_len) == 0) {
+    grpc_rb_compression_options_set_default_level_helper(
+        wrapper->wrapped, GRPC_COMPRESS_LEVEL_LOW);
+  } else if (strncmp(level_name, "medium", name_len) == 0) {
+    grpc_rb_compression_options_set_default_level_helper(
+        wrapper->wrapped, GRPC_COMPRESS_LEVEL_MED);
+  } else if (strncmp(level_name, "high", name_len) == 0) {
+    grpc_rb_compression_options_set_default_level_helper(
+        wrapper->wrapped, GRPC_COMPRESS_LEVEL_HIGH);
+  } else {
+    rb_raise(rb_eNameError,
+             "Invalid compression level name. Supported levels: none, low, "
+             "medium, high");
+  }
+
+  return Qnil;
+}
+
+/* Gets the internal value of a compression algorithm suitable as the value
+ * in a GRPC core channel arguments hash.
+ * Raises an error if the name of the algorithm passed in is invalid. */
+void grpc_rb_compression_options_get_internal_value_of_algorithm(
+    VALUE algorithm_name, grpc_compression_algorithm *compression_algorithm) {
+  VALUE ruby_str = Qnil;
+  char *name_str = NULL;
+  long name_len = 0;
+
+  /* Accept ruby symbol and string parameters. */
+  ruby_str = rb_funcall(algorithm_name, rb_intern("to_s"), 0);
+  name_str = RSTRING_PTR(ruby_str);
+  name_len = RSTRING_LEN(ruby_str);
+
+  /* Raise an error if the name isn't recognized as a compression algorithm by
+   * the algorithm parse function
+   * in GRPC core. */
+  if (!grpc_compression_algorithm_parse(name_str, name_len,
+                                        compression_algorithm)) {
+    rb_raise(rb_eNameError,
+             "Invalid compression algorithm name.");
+  }
+}
+
+/* Sets the default algorithm to the name of the algorithm passed in.
+ * Raises an error if the name is not a valid compression algorithm name. */
+VALUE grpc_rb_compression_options_set_default_algorithm(VALUE self,
+                                                        VALUE algorithm_name) {
+  grpc_rb_compression_options *wrapper = NULL;
+
+  TypedData_Get_Struct(self, grpc_rb_compression_options,
+                       &grpc_rb_compression_options_data_type, wrapper);
+
+  grpc_rb_compression_options_get_internal_value_of_algorithm(
+      algorithm_name, &wrapper->wrapped->default_algorithm.algorithm);
+  wrapper->wrapped->default_algorithm.is_set |= 1;
```

Comment:
This should really just be `= 1`. Using bitwise manipulation is more complicated and error-prone than necessary.